Code/Framework/AzCore/AzCore/Jobs/JobExecutor.cpp

```diff
@@ -0,0 +1,349 @@
+                        {
+                            TypeErasedJob* successor = job->m_graph->m_successors[job->m_successorOffset + j];
+                            if (--successor->m_dependencyCount == 0)
+                            {
+                                m_executor->Submit(*successor);
+                            }
+                        }
+
+                        job->m_graph->Release();
+                        --m_executor->m_remaining;
+
+                        job = m_queue.TryDequeue();
+                    }
+                }
+            }
+
+            AZStd::thread m_thread;
+            AZStd::atomic<bool> m_active;
+            AZStd::atomic<bool> m_busy;
+            AZStd::binary_semaphore m_semaphore;
+
+            ::AZ::JobExecutor* m_executor;
+            JobQueue m_queue;
+        };
+    } // namespace Internal
+
+    JobExecutor& JobExecutor::Instance()
+    {
+        // TODO: Create the default executor as part of a component (as in JobManagerComponent)
+        static JobExecutor executor;
+        return executor;
+    }
+
+    JobExecutor::JobExecutor(uint32_t threadCount)
+    {
+        // TODO: Configure thread count + affinity based on configuration
+        m_threadCount = threadCount == 0 ? AZStd::thread::hardware_concurrency() : threadCount;
+
+        m_workers = reinterpret_cast<Internal::JobWorker*>(azmalloc(m_threadCount * sizeof(Internal::JobWorker)));
+
+        bool affinitize = m_threadCount == AZStd::thread::hardware_concurrency();
+
+        AZStd::semaphore initSemaphore;
+
+        for (size_t i = 0; i != m_threadCount; ++i)
+        {
+            new (m_workers + i) Internal::JobWorker{};
+            m_workers[i].Spawn(*this, i, initSemaphore, affinitize);
+        }
+
+        for (size_t i = 0; i != m_threadCount; ++i)
+        {
+            initSemaphore.acquire();
+        }
+    }
+
+    JobExecutor::~JobExecutor()
+    {
+        for (size_t i = 0; i != m_threadCount; ++i)
+        {
+            m_workers[i].Join();
+            m_workers[i].~JobWorker();
+        }
+
+        azfree(m_workers);
+    }
+
+    void JobExecutor::Submit(Internal::CompiledJobGraph& graph)
+    {
+        // Submit all jobs that have no inbound edges
+        for (Internal::TypeErasedJob& job : graph.Jobs())
+        {
+            if (job.IsRoot())
+            {
+                Submit(job);
+            }
+        }
+    }
+
+    void JobExecutor::Submit(Internal::TypeErasedJob& job)
+    {
+        // TODO: Something more sophisticated is likely needed here.
+        // First, we are completely ignoring affinity.
+        // Second, some heuristics on core availability will help distribute work more effectively
+        ++m_remaining;
+        m_workers[++m_lastSubmission % m_threadCount].Enqueue(&job);
+    }
+
+    void JobExecutor::Drain()
+    {
+        while (m_remaining > 0)
+        {
+            AZStd::this_thread::sleep_for(AZStd::chrono::milliseconds{ 100 });
+        }
```

Comment:
If I understood correctly that `Drain` is expected to be called maybe one or more times per frame if a job chain is setup to include it, then this is going to be doing a bunch of busy waiting which wastes CPU cycles. Also 100 milliseconds seems like a very long time as it's a wait of several frames. Is there any way this can be updated with a signal that counts the number of remaining jobs instead?

Reply:
No this was just something I added during testing. I should swap it to be a signal yea